### ISSUE TITLE
ExtendedLog SSH, SFTP classes not working as expected

### ISSUE THREAD

**Issue by offsides** [proftpd/proftpd#1617]
Broken off from #1611, as this appears to be a genuine bug rather than a question/difference of philosophies.

### What I Did

I tried logging ALL,!SSH to get everything but CL_SSH messages, but it also suppressed CL_SFTP messages.

### What I Expected/Wanted

I was hoping to get the CL_SFTP log messages in my log file.  Based on playing with adding SFTP, SSH, !SFTP and !SSH to the ExtendedLog statements, it appears that the SFTP class has no impact on ExtendedLogs, and that CL_SFTP and CL_SSH are both included or excluded based on the SSH class.  Digging the code doesn't look like it should work that way, so I really am at a loss as to why it's doing so.

### ProFTPD Version and Configuration

Proftpd 1.3.8
```
### BEGIN DEVELOPMENT ONLE
LoadModule mod_ctrls_admin.c
ControlsEngine on
ControlsACLs all allow user root
AdminControlsEngine on
AdminControlsACLs all allow user root
### END DEVELOPMENT ONLE

# Load modules to extend ProFTPD.  Note that the order below MUST be kept!
LoadModule              mod_sql.c
LoadModule              mod_sql_mysql.c
LoadModule              mod_exec.c
LoadModule              mod_exec_mqueue.c
LoadModule              mod_sftp.c
LoadModule              mod_sftp_sql.c
LoadModule              mod_wrap2.c
LoadModule              mod_wrap2_sql.c
LoadModule              mod_unique_id.c
LoadModule              mod_ifsession.c

# FTP Server (base server settings)

# Basic server setup for a standalone server.
ServerType              standalone
# Port 21 is the standard FTP port.
Port                    21
# Specify the server IP.
DefaultAddress          <IP>
# Specify the name of the server.
ServerName              <FTP SERVERNAME>
# Disable IPv6
UseIPv6                 off
# Disable Reverse DNS lookups - we only care about the client IP address.
UseReverseDNS           off
# To prevent DoS attacks, set the maximum number of child processes
# to 30.  If you need to allow more than 30 concurrent connections
# at once, simply increase this value.  Note that this ONLY works
# in standalone mode, in inetd mode you should use an inetd server
# that allows you to limit maximum number of processes per service
# (such as xinetd).
MaxInstances            101

SystemLog		/var/log/proftpd/ftpserver.log
SyslogLevel		debug

# Global settings (FTP and SFTP)
<Global>
    DebugLevel		1
    Trace DEFAULT:10 ssh2:20 sftp:20 ifsession:20 auth.pam:20 sftp.sql:20
    TraceLog /var/log/proftpd/trace.log
    # Don't use GMT
    TimesGMT                off
    MaxClients              100
    
    # Security settings
    # Set the normal user and group permissions for the server.
    User                    ftp
    Group                   ftp
    # Drop root privs right away.
    RootRevoke              on
    # Show an empty server identification string for security purposes.
    ServerIdent             on "FTP Server ready."
    # Login banner
    DisplayConnect          /etc/ssh/login-banner
    # Chroot jail all users to their home directory.
    DefaultRoot             ~
    # Do not allow .ftpaccess files to override settings.
    AllowOverride           off
    # Umask 007 makes things group writable for automation purposes, and
    # prevents world access.
    Umask                   0007
    # Allow 60 seconds to authenticate after connection or the client is a
    # disconnected.
    TimeoutLogin            60
    # Allow 10 minutes of inactivity before a client is disconnected.
    TimeoutIdle             600
    TimeoutNoTransfer       600
    # Disconnect the user after a single failed password.
    MaxLoginAttempts        1
    # Disable the SITE CHGRP and SITE CHMOD commands, and active mode connections.
    <Limit SITE_CHGRP SITE_CHMOD PORT>
        DenyAll
    </Limit>
    # Allow files in /data/ftp to be overwritten.
    <Directory /data/ftp>
        AllowOverwrite  on
        PathDenyFilter  [\x01-\x1f\x7f"*:<>?\|]
    </Directory>
    # Allow uploads to be restarted.
    AllowStoreRestart       on
    
    # Global Logging
    # Custom log format similar to the NcFTPd logs.
    LogFormat               customf "%a %t F %u \"%r\" %s %b"
    LogFormat               customs "%a %t S %u \"%r\" %s %b"
    LogFormat               technical "%a %t %{UNIQUE_ID}e %u \"%r\" %s %b"
    LogFormat               eof "%a %t F %u %E"
    LogFormat               eos "%a %t S %u %E"
    LogFormat               techeos "%a %t %{UNIQUE_ID}e %u %E"
    # Log all commands, including EXIT, using the custom log format.
    
    # Authentication
    # wtmp logging is irrelevant as all FTP users are virtual.
    WtmpLog                 off
    # Do not allow PAM users for the same reasons.
    AuthPAM                 off
    # /etc/ftpusers is irrelevant as no system users are permitted.
    UseFtpUsers             off
    # Do not require users to have a valid shell because they are virtual.
    RequireValidShell       off
    # Use the SQL database as the only authentication source.
    AuthOrder               mod_sql.c
    
    # SQL parameters for mysql/mariadb.
    SQLBackend              mysql
    # Only use the SQL engine for authentication
    SQLEngine               auth
    # Log file for the SQL module.
    SQLLogFile              /var/log/proftpd/sql.log
    # SQL Connection info.
    SQLConnectInfo          <TABLE> <USER> <PASS>
    # SQL module options
    SQLOptions              IgnoreConfigFile
    # Use the system crypt(3) to check passwords.
    SQLAuthTypes            Crypt
    # Only do user lookups via SQL.
    SQLAuthenticate         users groups groupset
    # UID/GID for all virtual users (ftp/ftp).
    SQLDefaultUID           14
    SQLDefaultGID           50
    # Named queries used to determine what IP addresses a client is permitted
    # to connect from.  Used by mod_wrap2, below.
    SQLNamedQuery           get-allowed-clients FREEFORM "CALL allowedIPList('%u')"
    SQLNamedQuery           get-denied-clients SELECT "'ALL'"
    
    # IP-based access control using mod_wrap2.
    WrapEngine              on
    # Log file for the IP wrapper module.
    WrapLog                 /var/log/proftpd/wrap.log
    # Check all users for IP restrictions using the SQLNamedQuery definitions
    # above.
    WrapUserTables          * sql:/get-allowed-clients sql:/get-denied-clients
    # Messager sent to the client when they are allowed access from the
    # remote IP.
    WrapAllowMsg            "User '%u' allowed by access rules"
    # Message sent to the client when the user is denied access due to IP
    # restrictions.
    WrapDenyMsg             "User '%u' denied by access rules"
    
    # Execute system commands after specific FTP commands (event triggers).
    ExecEngine              on
    # Log file for the execution module.
    ExecLog                 /var/log/proftpd/exec.log
    # SysV IPC Message Queue key for triggers.
    ExecMqueueKey           0x78c0ffee
    # SysV IPC message type for triggers.
    ExecMqueueType          1
    # Trigger format (emulate NcFTPd's event format).
    ExecMqueueFormat        ncftpd
    # Timeout value if the message queue is full.
    ExecMqueueTimeout       5
    # Don't remove the message queue on exit.
    ExecMqueueCleanup       off
    # Successful login.
    ExecOnCommand           PASS mqueue: "LOGIN" "%U" "bogus" "%{time:%s}" "%v" "%a"
    # Failed login.
    ExecOnError             PASS mqueue: "FAILEDAUTH" "%U" "bogus" "%{time:%s}" "%v" "%a" "FTP"
    # Logout.
    ExecOnCommand           QUIT mqueue: "LOGOUT" "%U" "bogus" "%{time:%s}" "%v" "%a"
    # Successful file upload.
    ExecOnCommand           STOR mqueue: "UPLOAD" "%U" "bogus" "%{time:%s}" "%v" "%a" "%f" "%b" "bogus" "bogus" "bogus" "OK"
    # Failed file upload.
    ExecOnError             STOR mqueue: "UPLOAD" "%U" "bogus" "%{time:%s}" "%v" "%a" "%f" "%b" "bogus" "bogus" "bogus" "ERROR"
    # Successful file download.
    ExecOnCommand           RETR mqueue: "DOWNLOAD" "%U" "bogus" "%{time:%s}" "%v" "%a" "%f" "%b" "bogus" "bogus" "bogus" "OK"
    # Failed file download.
    ExecOnError             RETR mqueue: "DOWNLOAD" "%U" "bogus" "%{time:%s}" "%v" "%a" "%f" "%b" "bogus" "bogus" "bogus" "ERROR"
    # File rename.
    ExecOnCommand           RNTO mqueue: "RENAME" "%U" "bogus" "%{time:%s}" "%v" "%a" "%w" "X" "%f"
    
    # Define internal hosts for exclusion from some defensive measures
    <Class internal>
        From 127.0.0.1
        From 172.16.0.0/16
        From 172.25.0.0/16
        From 172.28.0.0/16
    </Class>
    
    # Don't use mod_delay for internal users
    <IfClass internal>
        DelayEngine off
    </IfClass>
    # Do use mod_delay for all other users
    <IfClass !internal>
        DelayEngine on
    </IfClass>

</Global>

# FTP Server-only settings
# User accounts table (view) and fields for user auth info.  NOTE: This view
# contains the necessary where clauses to determine if a user if active on the
# server.
SQLUserInfo             ftp_userinfo username password uid NULL homedir NULL
SQLGroupInfo            ftp_groupinfo groupname gid members
    ExtendedLog             /var/log/proftpd/proftpd.log ALL customf
    ExtendedLog             /var/log/proftpd/proftpd.log EXIT eof
    ExtendedLog             /var/log/proftpd/ftpdtechical.log ALL technical
    ExtendedLog             /var/log/proftpd/ftpdtechical.log EXIT techeos

# SFTP Server
<VirtualHost 0.0.0.0>
    # Specify the name of the server.
    ServerName              <SFTP SERVERNAME>
    # Port 22 is the standard SFTP port
    Port                    44
    # Enable mod_sftp for this virtualhost
    SFTPEngine              on
    SFTPLog                 /var/log/proftpd/sftp.log
    # Login banner
    SFTPDisplayBanner       /etc/ssh/login-banner
#    MaxLoginAttempts        5
#    ServerLog               /var/log/proftpd/ftpserver.log
    ExtendedLog             /var/log/proftpd/proftpd.log ALL customs
#    ExtendedLog             /var/log/proftpd/proftpd.log AUTH customs
    ExtendedLog             /var/log/proftpd/proftpd.log EXIT eos
    ExtendedLog             /var/log/proftpd/sftpdtechical.log ALL technical
    ExtendedLog             /var/log/proftpd/sftpdtechical.log EXIT techeos

    # ED448 Host Key
    SFTPHostKey             /data/hostkeys/ssh_host_ed448_key
    # ED25519 Host key
    SFTPHostKey             /data/hostkeys/ssh_host_ed25519_key
    # ECDSA Host Key
    SFTPHostKey             /data/hostkeys/ssh_host_ecdsa_key
    # RSA Host key
    SFTPHostKey             /data/hostkeys/ssh_host_rsa_key

    # Allow publickey and password auth by default
    SFTPAuthMethods         password publickey
    # Ignore setting times/permissions/attributes
    SFTPOptions             IgnoreSFTPUploadPerms IgnoreSFTPSetExtendedAttributes IgnoreSFTPSetOwners IgnoreSFTPSetPerms IgnoreSFTPSetTimes IgnoreSFTPUploadExtendedAttributes
    # Force rekeying after 1 hour or 2GB of data transfer
    SFTPRekey               required
    # Add periodic random data to SFTP sessions for protection against traffic analysis attacks
    SFTPTrafficPolicy       medium

    # List of supported Ciphers
    SFTPCiphers             aes256-gcm@openssh.com aes128-gcm@openssh.com aes256-ctr aes192-ctr aes128-ctr
    # List of supported MACs
    SFTPDigests             umac-128-etm@openssh.com hmac-sha2-512-etm@openssh.com hmac-sha2-256-etm@openssh.com umac-128@openssh.com hmac-sha2-512 hmac-sha2-256
    # List of supported Key Exchange protocols
    SFTPKeyExchanges        curve448-sha512 curve25519-sha256 ecdh-sha2-nistp521 ecdh-sha2-nistp384 ecdh-sha2-nistp256 diffie-hellman-group-exchange-sha256 diffie-hellman-group18-sha512 diffie-hellman-group16-sha512 diffie-hellman-group14-sha256
    # Require minimum 2048-bit RSA or 256-bit EC keys
    SFTPKeyLimits           MinimumRSASize 2048 MinimumECSize 256

    # User accounts table (view) and fields for user auth info.  NOTE: This view
    # contains the necessary where clauses to determine if a user if active on the
    # server.
    SQLUserInfo             sftp_userinfo username password uid gid homedir NULL
    SQLGroupInfo            sftp_groupinfo groupname gid members

    # Configure mod_ifsession to check for per-user/group settings before authentication
    IfSessionOptions        PerUnauthenticatedUser
    # If the user is in the group 'twofactor', require public key AND password auth
    <IfGroup twofactor>
        SFTPAuthMethods     publickey+password
    </IfGroup>

    # Named query to retrieve SSH keys for users
    SQLNamedQuery           get-user-authorized-keys SELECT "public_key FROM sshkeys WHERE username = '%U'"
    # Get all SSH keys from the database for SFTP access
    SFTPAuthorizedUserKeys  sql:/get-user-authorized-keys

</VirtualHost>
# vim: ts=4 sw=4 expandtab
```

**Comment by Castaglia** [comment]
There does indeed appear to have been a regression of [Bug#4067](http://bugs.proftpd.org/show_bug.cgi?id=4067) somewhere along the way.  Let's see what can be done to restore the intended functionality of the `SSH`, `SFTP` logging classes.

**Comment by Castaglia** [comment]
Looks like some of the changes for [Bug#4313](http://bugs.proftpd.org/show_bug.cgi?id=4313) are also involved in this regression.

**Comment by Castaglia** [comment]
https://github.com/proftpd/proftpd/pull/1621 should address this.  This fix will also affect some of the proposed documentation changes (_e.g._ https://github.com/proftpd/proftpd/pull/1620) as well, so I'll hold off on those doc changes until this logging issue settles out.

**Comment by Castaglia** [comment]
This should now be fixed in master, and in the 1.3.8 branch (via backport).  Thanks!
